libkbfs/tlf_journal.go

```diff
@@ -1219,27 +1271,77 @@ func (j *tlfJournal) getMDRange(
 	return j.mdJournal.getRange(start, stop)
 }
 
-func (j *tlfJournal) putMD(ctx context.Context, rmd *RootMetadata) (
-	MdID, error) {
+func (j *tlfJournal) doPutMD(ctx context.Context, rmd *RootMetadata,
+	irmd ImmutableRootMetadata, perRevMap unflushedPathsPerRevMap) (
+	MdID, bool, error) {
+	// Now take the lock and put the MD, merging in the unflushed
+	// paths while under the lock.
 	j.journalLock.Lock()
 	defer j.journalLock.Unlock()
 	if err := j.checkEnabledLocked(); err != nil {
-		return MdID{}, err
+		return MdID{}, false, err
+	}
+
+	if !j.unflushedPaths.appendToCache(irmd, perRevMap) {
+		return MdID{}, true, nil
 	}
+	// TODO: remove the revision from the cache on any errors below?
```

Comment:
Because we don't want to have the put succeed but need to retry the append due to a race, as we'll have to give up journal lock and leave things in an inconsistent state.